### PR TITLE
feat(dRICH): connect `irt` to `EICrecon` algorithms

### PR DESCRIPTION
**Pull request by c-dilks**
### Briefly, what does this PR introduce?
Initial connection of [`irt`](https://github.com/eic/irt) to `EICrecon` algorithms, via new algorithm `IrtCherenkovParticleID` and corresponding factory `IrtCherenkovParticleID_factory`.

#### Notes
- not production ready, therefore not enabled by default in `JEventProcessorPODIO`, but we have been using this for routine current dRICH performance studies
- difficult to unit test `IrtCherenkovParticleID`, because of the complexity of its inputs; prefer to defer testing to [DRICH `reconstruction_benchmarks`](https://eicweb.phy.anl.gov/EIC/benchmarks/reconstruction_benchmarks/-/merge_requests/293)
- `valgrind` leak checks: #709 (see below for comments)
- documentation: #708 

#### Caveats
- there are several known issues, which should be resolved in separate PRs
  - https://github.com/eic/EICrecon/issues/689
  - https://github.com/eic/EICrecon/issues/690
  - https://github.com/eic/EICrecon/issues/691
  - https://github.com/eic/EICrecon/issues/692
  - https://github.com/eic/EICrecon/issues/693
  - https://github.com/eic/EICrecon/issues/705
- `ROOT` dependence is needed, since `irt` heavily relies on objects such as `TVector3`
  - converters from `EDM4{hep,eic}` vectors to `ROOT` vectors are included in common header `Tools.h` (other downstream PID algorithms will use this header too)
- where possible, smart pointers are used in `IrtCherenkovParticleID`, but sometimes we had to unfortunately fall back to raw pointers, for compatibility with the `irt` repository, since:
  - `irt` does not use any smart pointers
  - `irt` classes typically include custom destructors; using smart pointers here in `EICrecon` runs the risk of 'double free' violations
  - any usage of the (banned) operator `new` includes a comment indicating where the allocated object will be destroyed (most likely within an `irt` class destructor)
  - `irt` objects can be streamed to ROOT files, and closing those ROOT files may `delete` the corresponding object(s); using smart pointers in this scenario can also cause 'double free' violations
  - we can try to fix this issue in separate PRs, but given the general fragility of streaming objects to ROOT files, I'm afraid this will not be straightforward
- not surprisingly, #709 identified a few leaks in the `irt` repository itself
  - [deferring fixing them until `irt` v2.0](https://github.com/eic/irt/issues/36), where we expect a [large refactor](https://github.com/eic/irt/pull/34)
  - fixed a few leaks from local geometry service `richgeo`; again unfortunately `irt` does not cooperate with smart pointers

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [x] Tests for the changes have been added:
  - https://eicweb.phy.anl.gov/EIC/benchmarks/reconstruction_benchmarks/-/merge_requests/293
- [x] Documentation has been added / updated
  - #708
- [x] Changes have been communicated to collaborators
  - @chchatte92 
  - @alexander-kiselev
  - @DelloStritto

### Does this PR introduce breaking changes? What changes might users need to make to their code?
no

### Does this PR change default behavior?
no

### CODE REVIEW

File: src/algorithms/pid/IrtCherenkovParticleID.cc
Comment (veprbl) [inline src/algorithms/pid/IrtCherenkovParticleID.cc:28]: This case clearly not tested, followed up by dereferencing `detectors.begin()`. Please avoid adding dead code.

File: src/algorithms/pid/Tools.h
Comment (veprbl) [inline src/algorithms/pid/Tools.h:94]: Somehow, all trivial functions ended up having (redundant) docstrings, but these two unobvious functions don't have explanation for their purpose

File: src/services/geometry/richgeo/IrtGeoPFRICH.h
Comment (c-dilks) [reply src/services/geometry/richgeo/IrtGeoPFRICH.h]: `IRT` class objects may sometimes assume ownership of these raw pointers, copy them, move them etc.. I haven't tracked down precisely what the issue is, but let me see if at least some of these can be made smart.

File: src/algorithms/pid/IrtCherenkovParticleID.cc
Comment (c-dilks) [reply src/algorithms/pid/IrtCherenkovParticleID.cc:92]: I would prefer a thread-safe PDG service, even if using `TDatabasePDG` is restricted to initialization in this case. If I switch this back to `TDatabasePDG`, I have to finish #416, and even though that's not much work, it doesn't seem worth the effort when a thread safe service (that could read `TDatabasePDG` on initialization) is a much better approach for all.

File: src/services/geometry/richgeo/IrtGeoPFRICH.h
Comment (c-dilks) [reply src/services/geometry/richgeo/IrtGeoPFRICH.h]: These are the only ones that still work as a smart pointer:
```cpp
      std::shared_ptr<FlatSurface>             m_surfEntrance;
      std::shared_ptr<CherenkovPhotonDetector> m_irtPhotonDetector;
      std::shared_ptr<FlatSurface>             m_aerogelFlatSurface;
      std::shared_ptr<FlatSurface>             m_filterFlatSurface;
```
Converting the others to smart pointers breaks ROOT streamability; here is a stacktrace from [drich-dev auxfile creation](https://github.com/eic/drich-dev/blob/main/src/create_irt_auxfile.cpp), which uses these `IrtGeo`-derived classes to stream objects to ROOT files (not needed for EICrecon, but some other dRICH devs prefer to have this feature):
```
===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================
#0  0x00007f53f471e9d3 in wait4 () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007f53f469719b in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x00007f53f539394d in TUnixSystem::StackTrace() () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libCore.so.6.26
#3  0x00007f53f53932e4 in TUnixSystem::DispatchSignals(ESignals) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libCore.so.6.26
#4  <signal handler called>
#5  0x00007f53f4a57d10 in vtable for __cxxabiv1::__vmi_class_type_info () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007f53f533031a in TClass::Store(TBuffer&) const () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libCore.so.6.26
#7  0x00007f53f4d92a31 in TBufferFile::WriteClass(TClass const*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#8  0x00007f53f4d936d1 in TBufferFile::WriteObjectClass(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#9  0x00007f53f4d9affb in TBufferIO::WriteObjectAny(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#10 0x00007f53f4d8e9ec in TBufferFile::WriteFastArray(void**, TClass const*, int, bool, TMemberStreamer*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#11 0x00007f53f4fff0be in int TStreamerInfo::WriteBufferAux<char**>(TBuffer&, char** const&, TStreamerInfo::TCompInfo* const*, int, int, int, int, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#12 0x00007f53f4e6123d in TStreamerInfoActions::GenericWriteAction(TBuffer&, void*, TStreamerInfoActions::TConfiguration const*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#13 0x00007f53f4d945b6 in TBufferFile::WriteClassBuffer(TClass const*, void*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#14 0x00007f53f428261d in OpticalBoundary::Streamer (this=0x564ef2fe03e0, R__b=...) at /home/dilks/d/drich-dev/irt/build/G__IRT.cxx:2340
#15 0x00007f53f4d9370f in TBufferFile::WriteObjectClass(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#16 0x00007f53f4d9aeeb in TBufferIO::WriteObjectAny(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#17 0x00007f53f4e13711 in TGenCollectionStreamer::WriteObjects(int, TBuffer&) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#18 0x00007f53f4de27f3 in TCollectionStreamer::Streamer(TBuffer&, void*, int, TClass*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#19 0x00007f53f4d8e375 in TBufferFile::WriteFastArray(void*, TClass const*, int, TMemberStreamer*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#20 0x00007f53f500189a in int TStreamerInfo::WriteBufferAux<char**>(TBuffer&, char** const&, TStreamerInfo::TCompInfo* const*, int, int, int, int, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#21 0x00007f53f4e6123d in TStreamerInfoActions::GenericWriteAction(TBuffer&, void*, TStreamerInfoActions::TConfiguration const*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#22 0x00007f53f4d945b6 in TBufferFile::WriteClassBuffer(TClass const*, void*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#23 0x00007f53f4282e5d in CherenkovPhotonDetector::Streamer (this=0x564ef2fc8f00, R__b=...) at /home/dilks/d/drich-dev/irt/build/G__IRT.cxx:2406
#24 0x00007f53f4d9370f in TBufferFile::WriteObjectClass(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#25 0x00007f53f4d9aeeb in TBufferIO::WriteObjectAny(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#26 0x00007f53f4e13711 in TGenCollectionStreamer::WriteObjects(int, TBuffer&) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#27 0x00007f53f4de27f3 in TCollectionStreamer::Streamer(TBuffer&, void*, int, TClass*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#28 0x00007f53f4d8e375 in TBufferFile::WriteFastArray(void*, TClass const*, int, TMemberStreamer*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#29 0x00007f53f500189a in int TStreamerInfo::WriteBufferAux<char**>(TBuffer&, char** const&, TStreamerInfo::TCompInfo* const*, int, int, int, int, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#30 0x00007f53f4e6123d in TStreamerInfoActions::GenericWriteAction(TBuffer&, void*, TStreamerInfoActions::TConfiguration const*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#31 0x00007f53f4d945b6 in TBufferFile::WriteClassBuffer(TClass const*, void*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#32 0x00007f53f4285209 in CherenkovDetector::Streamer (this=0x564ed4553950, R__b=...) at /home/dilks/d/drich-dev/irt/build/G__IRT.cxx:2696
#33 0x00007f53f4d9370f in TBufferFile::WriteObjectClass(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#34 0x00007f53f4d9aeeb in TBufferIO::WriteObjectAny(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#35 0x00007f53f4d8e9ec in TBufferFile::WriteFastArray(void**, TClass const*, int, bool, TMemberStreamer*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#36 0x00007f53f4ff8922 in int TStreamerInfo::WriteBufferAux<TVirtualCollectionProxy>(TBuffer&, TVirtualCollectionProxy const&, TStreamerInfo::TCompInfo* const*, int, int, int, int, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#37 0x00007f53f4e94d0b in TStreamerInfoActions::GenericLooper::GenericWrite(TBuffer&, void*, void const*, TStreamerInfoActions::TLoopConfiguration const*, TStreamerInfoActions::TConfiguration const*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#38 0x00007f53f4d8d20c in TBufferFile::ApplySequence(TStreamerInfoActions::TActionSequence const&, void*, void*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#39 0x00007f53f5001a5c in int TStreamerInfo::WriteBufferAux<char**>(TBuffer&, char** const&, TStreamerInfo::TCompInfo* const*, int, int, int, int, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#40 0x00007f53f4e6123d in TStreamerInfoActions::GenericWriteAction(TBuffer&, void*, TStreamerInfoActions::TConfiguration const*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#41 0x00007f53f4d945b6 in TBufferFile::WriteClassBuffer(TClass const*, void*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#42 0x00007f53f428565b in CherenkovDetectorCollection::Streamer (this=0x564ed470c7a0, R__b=...) at /home/dilks/d/drich-dev/irt/build/G__IRT.cxx:2729
#43 0x00007f53f4e3a3f6 in TKey::TKey(TObject const*, char const*, int, TDirectory*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#44 0x00007f53f4dfa245 in TFile::CreateKey(TDirectory*, TObject const*, char const*, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#45 0x00007f53f4deb66e in TDirectoryFile::WriteTObject(TObject const*, char const*, char const*, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#46 0x00007f53f5275649 in TObject::Write(char const*, int, int) const () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libCore.so.6.26
#47 0x0000564ece45165a in main (argc=2, argv=0x7fff5edb6d18) at src/create_irt_auxfile.cpp:63
===========================================================
```

Comment (c-dilks) [reply src/services/geometry/richgeo/IrtGeoPFRICH.h]: Do you know of a way to have ROOT I/O play nicely with streaming objects owned by smart pointers?

**Comment by c-dilks** [comment]
> it costed me quite some time to tune a right combination of pointers and TRef's in the binary ROOT dump describing the IRT optics. I'd really appreciate if none of this gets touched unless absolutely needed. Feel free to add any "//!" fields of course.

Very much agree. I have not changed any of this in `IRT` and do not plan to. The issue under discussion here is the preference toward smart pointers in EICrecon, and getting smart pointers to play nicely with raw pointers and `TRef`s is not for the faint of heart and has also costed me much time. I have therefore used raw pointers in EICrecon as needed, and smart pointers only when it's clear `IRT` will not mess with them.

We may have some memory leaks in `IRT` itself though; it would be a good idea to check all destructors to make sure any `new` has a `delete`. See https://github.com/eic/irt/issues/36

Perhaps we wait until after merging your `pfrich` branch to deal with these leaks.

**Comment by alexander-kiselev** [comment]
Hi Chris,

  concerning the leaks, this is my usual problem, admittedly. When in a rush, I do not care about destructors at all, but then prefer to use stack variables whenever possible. The latter is a way asking for troubles with ROOT though. So yes, will need to check all the classes that they have proper ~ methods, at some point.

  Cheers,
    Alexander.


________________________________

File: src/services/geometry/richgeo/IrtGeoPFRICH.h
Comment (veprbl) [reply src/services/geometry/richgeo/IrtGeoPFRICH.h]: New comment is more specific, thanks.

> `IRT` class objects may sometimes assume ownership of these raw pointers, copy them, move them etc.. I haven't tracked down precisely what the issue is, but let me see if at least some of these can be made smart.

But you still call `operator delete *` on the objects in the end, it should not make a difference if you do it twice with a smart pointer or manually.

Comment (c-dilks) [reply src/services/geometry/richgeo/IrtGeoPFRICH.h]: Based on comments from @alexander-kiselev below, and on private discussions we've had in the past, I think it is safer to just use these raw pointers here. I tried my best to use smart pointers everywhere in the algorithm too, `IrtCherenkovParticleID`, but again passing `std::shared_ptr<T>::get` to IRT methods tends to cause unexpected results or seg faults.

As long as we confirm that any `operator new *` has a corresponding `operator delete *`, we should be okay. There are [leaks in IRT, however](https://github.com/eic/irt/issues/36), where some destructors need implementation. Since we don't intend for this IRT to be run in production yet, these leaks should not block this PR's approval (but should be addressed ASAP).

File: src/global/pid/IrtCherenkovParticleID_factory.cc
Comment (c-dilks) [reply src/global/pid/IrtCherenkovParticleID_factory.cc:92]: Thanks, I forgot to do this on our IRT development branch too.

File: src/algorithms/pid/IrtCherenkovParticleID.cc
Comment (c-dilks) [reply src/algorithms/pid/IrtCherenkovParticleID.cc:5]: `MergeTracksPID` has unit tests, `IrtCherenkovParticleID` does not. Neither are called by default `eicrecon` execution. I can't think of a unit test for `IrtCherenkovParticleID` that would be stable against changes in geometry.

File: src/algorithms/pid/IrtCherenkovParticleID.cc
Comment (c-dilks) [reply src/algorithms/pid/IrtCherenkovParticleID.cc:28]: These PID algorithms are designed to be callable from outside of EICrecon too. @alexander-kiselev prefers to store the IRT geometry, a `CherenkovDetectorCollection` object, in a ROOT file, and we want to maintain compatibility with that workflow here as much as possible. Checking `detectors.size()` is a useful sanity check when reading a ROOT file that contains a `CherenkovDetectorCollection`. Since this check is in `AlgorithmInit`, there is no performance loss.

Here in EICrecon, `CherenkovDetectorCollection` objects are generated by the `richgeo` service, where we guarantee `detectors.size()==1`; we do not generate such ROOT files, although we [maintain a program to do this](https://github.com/eic/drich-dev/blob/269bf3ceb1616dcf78aa5449262d4fc45f1b41b6/src/create_irt_auxfile.cpp#L1) for anyone who wants this feature.

So this code is only dead from the point of view of EICrecon usage.

File: src/algorithms/pid/Tools.h
Comment (c-dilks) [reply src/algorithms/pid/Tools.h:94]: I documented the new ones I wrote, but forgot to document the ones directly copied from Juggler. Will fix.

File: src/algorithms/pid/IrtCherenkovParticleID.cc
Comment (veprbl) [reply src/algorithms/pid/IrtCherenkovParticleID.cc:28]: If `detectors.size() == 0`, it will crash on `auto this_detector = *detectors.begin();`, value of `m_init_failed` will never be used.

File: src/algorithms/pid/IrtCherenkovParticleID.cc
Comment (veprbl) [reply src/algorithms/pid/IrtCherenkovParticleID.cc:5]: I assumed, this could be tested by providing some mock geometry. Otherwise, it might be difficult. I haven't figured out how we could mock geometry to write cases for code that uses DD4hep.

**Comment by alexander-kiselev** [comment]
I should be able to work on this past the PID review in July. A showstopper here can be a necessity to fish out the optical element details out of the geometry service, which has no clue about the reflective and refractive surface sequence in the object hierarchy. Let me mention that the default pfRICH optics is now way more complicated than the dRICH one. Importing a binary .root file with such an optical configuration created along with the pfRICH geometry creation (be that in a standalone GEANT or eventually in a native TGeo code) is of course a natural choice, but somehow the software group never liked the idea of creating a dependency on IRT library in the reconstruction codes. Can this please be agreed upon, perhaps as an exception?
________________________________

**Comment by c-dilks** [comment]
> I should be able to work on this past the PID review in July. A showstopper here can be a necessity to fish out the optical element details out of the geometry service, which has no clue about the reflective and refractive surface sequence in the object hierarchy. Let me mention that the default pfRICH optics is now way more complicated than the dRICH one. Importing a binary .root file with such an optical configuration created along with the pfRICH geometry creation (be that in a standalone GEANT or eventually in a native TGeo code) is of course a natural choice, but somehow the software group never liked the idea of creating a dependency on IRT library in the reconstruction codes. Can this please be agreed upon, perhaps as an exception?

This is why we still support this capability here in reconstruction. See the signature here:
https://github.com/eic/EICrecon/blob/de1a1a28d0486a8366af082083d00fb7114c36c6/src/algorithms/pid/IrtCherenkovParticleID.h#L35-L38

`AlgorithmInit` is called once at the beginning (before any events, to configure the algorithm), taking a `CherenkovDetectorCollection*`, which in EICrecon comes from the geometry service. It could also come from anywhere else, in particular, a ROOT file that contains a `CherenkovDetectorCollection` object.

One can overload `AlgorithmInit` to add a ROOT file reader:
```cpp
void AlgorithmInit(TFile* irt_root_file, std::shared_ptr<spdlog::logger>& logger) {
  AlgorithmInit(irt_root_file->Get<CherenkovDetectorCollection>("NameOfCherenkovDetectorCollection", logger);
}
```

This `IrtCherenkovParticleID` class does not care about the origin; it configures IRT and runs it. It has no dependence on EICrecon itself or on JANA2, despite existing in the EICrecon repository; it is a part of the "independent" algorithms that will eventually be fully separated from the framework that runs them, EICrecon+JANA2.

`IrtCherenkovParticleID` algorithm does depend on IRT (think of `IrtCherenkovParticleID` as a sort of wrapper for IRT that is aware of the data model). The `epic` geometry code does not depend on IRT, but I leave it to you to argue a case for such dependence; however, this PR review is not the appropriate place for that argument.

File: src/algorithms/pid/IrtCherenkovParticleID.cc
Comment (c-dilks) [reply src/algorithms/pid/IrtCherenkovParticleID.cc:5]: Mock DD4hep geometry would be nice. It might be more effort than worth it though, if we can trust the downstream benchmarks enough, at least in this case. Unit testing should be added to `irt` itself, since `IrtCherenkovParticleID` here is effectively a wrapper of `irt`.

**Comment by alexander-kiselev** [comment]
sounds good. I recall in the ATHENA times (but also in the early ePIC times) there were objections against making reconstruction codes explicitly dependent on IRT optics description (recall front door vs back door import in Juggler). Perhaps I'm mixing up the topics though.
________________________________

File: src/algorithms/pid/IrtCherenkovParticleID.cc
Comment (c-dilks) [reply src/algorithms/pid/IrtCherenkovParticleID.cc]: Fixed: [c29fcbe](https://github.com/eic/EICrecon/pull/707/commits/c29fcbe452cbe54702c05663ce67e30d61d3fc7b)

File: src/algorithms/pid/IrtCherenkovParticleID.cc
Comment (c-dilks) [reply src/algorithms/pid/IrtCherenkovParticleID.cc:5]: I'm not the `irt` maintainer, but at least we can dream:
https://github.com/eic/irt/issues/37

**Comment by c-dilks** [comment]
@veprbl all conversations are resolved. Can you please review, when you get a chance?

File: src/global/pid/IrtCherenkovParticleID_factory.cc
Comment (veprbl) [inline src/global/pid/IrtCherenkovParticleID_factory.cc:55]:
```suggestion
    auto input_tag   = GetInputTags().at(tag_num++);
```

File: src/global/pid/IrtCherenkovParticleID_factory.cc
Comment (veprbl) [inline src/global/pid/IrtCherenkovParticleID_factory.cc:59]: This place looks a bit strange. You parse the name and then seemingly encode it again.

**Review by veprbl — DISMISSED** [review]

File: src/global/pid/IrtCherenkovParticleID_factory.cc
Comment (c-dilks) [reply src/global/pid/IrtCherenkovParticleID_factory.cc:59]: `richgeo::ParseRadiatorName` searches for the substring `"Aerogel"` or `"Gas"` (somewhat case insensitive) in `input_tag` (e.g., `"DRICHAerogelTracks"`) and returns the corresponding `int` radiator ID. `richgeo::RadiatorName` converts the radiator ID to the radiator name, either `"Aerogel"` or `"Gas"`, a `string` radiator ID.

Unfortunately, some user code uses the `int` ID and others the `string` ID.

Comment (veprbl) [reply src/global/pid/IrtCherenkovParticleID_factory.cc:59]: Perhaps, `richgeo::ParseRadiatorName` could be named `richgeo::ParseTag`.

**Review by veprbl — APPROVED** [review]

**Pull request merged**
